openconnect-gp-okta

```diff
@@ -318,16 +343,51 @@ def main(
     if password is None:
         password = click.prompt('Password', hide_input=True)
 
+    cert_crt = cert_key = None
+    cert_args = []
+    if cert_path is not None:
+        if cert_password_cmd is not None:
+            cert_password = run_cmd(cert_password_cmd, "Cert password")
+        if cert_password is None:
+            cert_password = click.prompt('Cert Password', hide_input=True)
+
+        cert_args = ['-c', cert_path, '-p', cert_password]
+
+        if cert_path.endswith('.p12'):
+            with open(cert_path, 'rb') as fp:
+                cert_key, cert_crt, _ = pkcs12.load_key_and_certificates(
+                    fp.read(),
+                    cert_password.encode('utf-8'))
+
     factor_priorities = {
         'token:software:totp': 0 if totp_key is None else 2,
         'push': 1,
         **dict(factor_priorities)}
 
     with requests.Session() as s:
-        s.mount('https://', TLSAdapter())
-        saml_req_url = prelogin(s, gateway)
-        saml_resp_url, saml_resp_data = okta_saml(s, saml_req_url, username, password, factor_priorities, totp_key)
-        saml_username, prelogin_cookie = complete_saml(s, saml_resp_url, saml_resp_data)
+        if cert_crt is not None and cert_key is not None:
+            crt = cert_crt.public_bytes(encoding=Encoding.PEM)
+            key = cert_key.private_bytes(encoding=Encoding.PEM,
+                                         format=PrivateFormat.PKCS8,
+                                         encryption_algorithm=NoEncryption())
+
+            # This is a kludge because there's no interface to take an
+            # in-memory certificate and key.
+            with tempfile.TemporaryDirectory() as tdir:
+                with tempfile.NamedTemporaryFile(dir=tdir, prefix='crt-', buffering=0) as crt_file, \
+                     tempfile.NamedTemporaryFile(dir=tdir, prefix='key-', buffering=0) as key_file:
+                    crt_file.write(crt)
+                    key_file.write(key)
+                    s.cert = (crt_file.name, key_file.name)
+                    s.mount('https://', TLSAdapter())
+                    saml_req_url = prelogin(s, gateway)
+                    saml_resp_url, saml_resp_data = okta_saml(s, saml_req_url, username, password, factor_priorities, totp_key)
+                    saml_username, prelogin_cookie = complete_saml(s, saml_resp_url, saml_resp_data)
+        else:
+                s.mount('https://', TLSAdapter())
```

Comment:
nit: weird indentation?

Also, this logic seems to be duplicated. Why do we even need the else branch?

Reply:
In the first case, we wan't the TemporaryDirectory to be cleaned up. The else doesn't need the temporary files.  The common bits could be factored into a function.

Reply:
Done.